src/control/shaperecognizer/ShapeRecognizer.cpp

```diff
@@ -297,22 +297,33 @@ auto ShapeRecognizer::recognizePatterns(Stroke* stroke) -> ShapeRecognizerResult
 
         if (n == 1)  // current stroke is a line
         {
+            bool aligned = false;
             if (fabs(rs->angle) < SLANT_TOLERANCE)  // nearly horizontal
             {
                 rs->angle = 0.0;
                 rs->y1 = rs->y2 = rs->ycenter;
+                aligned = true;
             }
             if (fabs(rs->angle) > M_PI / 2 - SLANT_TOLERANCE)  // nearly vertical
             {
                 rs->angle = (rs->angle > 0) ? (M_PI / 2) : (-M_PI / 2);
                 rs->x1 = rs->x2 = rs->xcenter;
+                aligned = true;
             }
 
             auto* s = new Stroke();
             s->applyStyleFrom(this->stroke);
 
-            s->addPoint(Point(rs->x1, rs->y1));
-            s->addPoint(Point(rs->x2, rs->y2));
+            if (aligned)
+            {
+                s->addPoint(Point(rs->x1, rs->y1));
+                s->addPoint(Point(rs->x2, rs->y2));
+            } else {
+                s->addPoint(Point(stroke->getPoints()[0].x, stroke->getPoints()[0].y));
+                s->addPoint(Point(stroke->getPoints()[stroke->getPointCount() - 1].x,
+                                  stroke->getPoints()[stroke->getPointCount() - 1].y));
```

Comment:
Please store the vector returned from `stroke->getPoints()` in a local variable before creating the points using this local vector. Doing it this way makes it hard to read and also forces several derefenciations of pointers that cost a lot of time.